test/test_ForwardDiff.jl

```diff
@@ -0,0 +1,221 @@
+cost02_with_dt(rand())
+FDgrad(cost02, rand())
+FDgrad(cost02_with_dt, rand())
+fin_diff(cost02, rand())
+### test vs finite difference
+@test all([test_vs_fin_diff(cost02_with_dt, q; atol=1e-7) for q=vcat(0,π,rand(tests_repetition)*2π)])
+### check that we still get NaN's
+#### In this case, it seems that if sin(p) is small, we don't get a NaN
+@test_broken all(.!isnan.(FDgrad.(cost02, range(π/2,tests_repetition))))
+
+## test vs ForwardDiff on DiffEq
+function cost02_via_DiffEq(par; kwargs...)
+    op = create(ba0)+destroy(ba0)
+    schrod(u,p,_) = -im*sin(p)*(op.data*u)
+    prob = ODEProblem(schrod, psi.data, (0.0, 1.0), par; dtmax=exp2(-4), saveat=(0.0, 1.0), abstol=1e-5, reltol=1e-5, alg=Tsit5(), kwargs...)
+    sol = solve(prob)
+    abs2(target0.data'*last(sol.u))
+end
+### check that nothing fails
+cost02_via_DiffEq(rand())
+FDgrad(cost02_via_DiffEq, rand())
+@assert all([(p=2π*rand(); isapprox(cost02_via_DiffEq(p), cost02(p); atol=1e-9)) for _=1:tests_repetition])
+### test vs DiffEq.jl
+@test let
+    p = 2π*rand(tests_repetition)
+    gde = FDgrad.(cost02_via_DiffEq, p)
+    gqo = FDgrad.(cost02, p)
+    NaN_check = isnan.(gqo) == isnan.(gde) # have NaN at same places
+    if !NaN_check
+        return NaN_check
+    end
+    val_check = isapprox(filter(!isnan, gqo), filter(!isnan, gde), atol=1e-12)
+    val_check && NaN_check
+end
+### check that we still get NaN's
+@test_broken all(.!isnan.(FDgrad.(cost02_via_DiffEq, range(π/2,tests_repetition))))
+
+# ex1
+## 3 level kerr transmon with drive
+ba1 = FockBasis(2)
+T2 = (1+rand())*1e4
+ω00, α0 = 0.1randn(), -0.2+0.05rand()
+function get_Ht(p::Vector{<:Tp}) where Tp
+    ω0, α = Tp(ω00), Tp(α0)
+    A, freq, ϕ, T = p
+    op = 2π*([number(ba1), 2\create(ba1)*number(ba1)*destroy(ba1), im*(create(ba1)-destroy(ba1))])
+    fω(t) = ω0
+    fα(t) = α
+    fΩ(t) = A*cospi(2t*freq + 2ϕ)*sinpi(t/T)^2
+    H_at_t = LazySum(zeros(Tp,length(op)), op)
+    function Ht(t,_)
+        H_at_t.factors.= (fω(t), fα(t), fΩ(t))
+        return H_at_t
+    end
+    return Ht
+end
+## initial states
+ψ01 = Operator(SpinBasis(1/2), basisstate(ba1, 1), basisstate(ba1, 2))
+## target states
+target1 = ψ01*exp(im*0.5π*dense(sigmax(SpinBasis(1/2)))) # x gate
+## cost function using QO.jl
+function cost1(par; kwargs...)
+    T = par[4]
+    Ht = get_Ht(par)
+    ts = (0.0, T)
+    _, ψT = timeevolution.schroedinger_dynamic(ts, ψ01, Ht; abstol=1e-9, reltol=1e-9, dtmax=exp2(-5), kwargs...)
+    1-abs2(tr(target1'last(ψT))/2)*exp(-T/T2)
+end
+cost1_with_dt(par; kwargs...) = cost1(par; dt=exp2(-5), kwargs...)
+
+p0 = [0.3, ω00, 0.25, 10.0]
+rp(k) = p0 .* ( ones(length(p0))+k*(1 .-2rand(length(p0))) )
+### check that nothing fails
+rp(rand())
+cost1(p0)
+FDgrad(cost1, p0)
+FDgrad(cost1_with_dt, p0)
+### test vs finite difference
+#### there somtimes an issue with the gradient at p0, which is alot less accurate
+@test all([test_vs_fin_diff(cost1_with_dt, p; atol=1e-6) for p=rp.(range(1e-9, 0.1, tests_repetition))])
```

Comment:
There is a small unstable region around `p0`, where AD and finite diff differ by `1e-2`. Not sure why that is.

Reply:
Would just removing that test make sense? Is it really testing anything that the previous set of much simpler tests do not? This is about testing the autodiff, not about a study of numerical stability.

Reply:
Yes.
I didn't check this example vs DiffEq, but given that the other examples I compared showed less than 1e-12 difference I think we can remove this test.
And also the 1'st test?

Reply:
I am not sure which one you refer to as 1st. Testing both the static and dynamic Hamiltonians seems valuable to me.

Reply:
Sure

Reply:
Keeping the two tests you specifically cited below seems reasonable to me. You can also guide yourself by the coverage report - we should aim to have all new lines covered by tests.